pkg/gateway/model_build_access_log_subscription.go

```diff
@@ -0,0 +1,86 @@
+package gateway
+
+import (
+	"context"
+	"fmt"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	anv1alpha1 "github.com/aws/aws-application-networking-k8s/pkg/apis/applicationnetworking/v1alpha1"
+	"github.com/aws/aws-application-networking-k8s/pkg/k8s"
+	"github.com/aws/aws-application-networking-k8s/pkg/model/core"
+	model "github.com/aws/aws-application-networking-k8s/pkg/model/lattice"
+	"github.com/aws/aws-application-networking-k8s/pkg/utils/gwlog"
+)
+
+type AccessLogSubscriptionModelBuilder interface {
+	Build(ctx context.Context, alp *anv1alpha1.AccessLogPolicy) (core.Stack, *model.AccessLogSubscription, error)
+}
+
+type accessLogSubscriptionModelBuilder struct {
+	log    gwlog.Logger
+	client client.Client
+}
+
+func NewAccessLogSubscriptionModelBuilder(log gwlog.Logger, client client.Client) *accessLogSubscriptionModelBuilder {
+	return &accessLogSubscriptionModelBuilder{
+		client: client,
+		log:    log,
+	}
+}
+
+func (b *accessLogSubscriptionModelBuilder) Build(
+	ctx context.Context,
+	accessLogPolicy *anv1alpha1.AccessLogPolicy,
+) (core.Stack, *model.AccessLogSubscription, error) {
+	stack := core.NewDefaultStack(core.StackID(k8s.NamespacedName(accessLogPolicy)))
+
+	task := accessLogSubscriptionModelBuildTask{
+		stack:           stack,
+		accessLogPolicy: accessLogPolicy,
+	}
+
+	if err := task.run(ctx); err != nil {
+		return nil, nil, err
+	}
+
+	return task.stack, task.accessLogSubscription, nil
+}
+
+type accessLogSubscriptionModelBuildTask struct {
+	stack                 core.Stack
+	accessLogPolicy       *anv1alpha1.AccessLogPolicy
+	accessLogSubscription *model.AccessLogSubscription
+}
+
+func (t *accessLogSubscriptionModelBuildTask) run(ctx context.Context) error {
+	sourceType := model.ServiceSourceType
```

Comment:
One general comment, since we are doing a brand new e2e code path for `accessLogPolicy`, can we don't follow the old side-effect style code? I think old style is a little bit hard to trace the data flow (tho we already got used to it)

Can we try to use stateless functions as much as we can?

 for example:
`accessLogSubscriptionModelBuildTask` could don't have member variables `stack`, `accessLogPolicy ` , `accessLogSubscription`. In stead, use function input params and return vaule to represent them:
`func (t *accessLogSubscriptionModelBuildTask) run(ctx context.Context, accessLogPolicy       *anv1alpha1.AccessLogPolicy ) *model.AccessLogSubscription, error `
(And I think either `stack` or  `accessLogSubscription` just need one, don't need both)

Another example is `(s *accessLogSubscriptionSynthesizer) Synthesize()`, can we removed member variable `stack` of accessLogSubscriptionSynthesizer and put it in the input parameter?

`func (s *accessLogSubscriptionSynthesizer) Synthesize(ctx context.Context, accessLogSubscription *model.AccessLogSubscription) error `

Reply:
While I agree a functional code path is ideal, I think we are at a point now where deviation from the existing style would require a much more substantial refactor of the codebase. I would rather keep things consistent and have predictable side-effects, than have multiple different controllers with wildly different styles of code

Perhaps we should create a refactoring ticket to migrate over if the existing code style is causing churn?

Reply:
Ok, make sense to do code style refactoring in a seperate PR, I can create a issue for it